### ISSUE TITLE
[PATCH] twisted.spread.util.FilePager doesn't callback when sending an empty file

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3023]
|<img alt="zectbumo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| zectbumo reported|
|-|-|
|Trac ID|trac#3023|
|Type|defect|
|Created|2008-02-07 03:34:24Z|
|Branch|https://github.com/twisted/twisted/tree/filepager-empty-file-3023|

If you use FilePager to send an empty file then the completion callback never gets called.

Attachments:

* [util.py.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9c7/9c700f06a78a028f49144bd1c4efcf6f7a830f8f/37827fffd60456cb8dbd7217b4a392ac2d84fea1.patch) (821 bytes) - added by zectbumo on 2008-02-07 03:39:50Z - 
* [util.py.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9c7/9c700f06a78a028f49144bd1c4efcf6f7a830f8f/35b5971e78061a1595abcf39172ca579a4a86345.patch) (546 bytes) - added by zectbumo on 2008-02-07 19:46:58Z - this also seems to fix it
* [test_pb.py.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9c7/9c700f06a78a028f49144bd1c4efcf6f7a830f8f/5a915f9c87c81eb5a58671e333a8624182341f0b.patch) (1227 bytes) - added by zectbumo on 2008-02-09 18:22:50Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3023 3023
type__defect defect
reporter__zectbumo zectbumo
priority__highest highest
milestone__ 
branch__branches_filepager_empty_file_3023 branches/filepager-empty-file-3023
branch_author__ 
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__ 
time__1202355264000000 1202355264000000
changetime__1203152420000000 1203152420000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="zectbumo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|zectbumo commented|
|-|-|

I like patch 2 better since it is more straight forward

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to zectbumo|
|-|-|

Thanks for the patch. Some comments:
 * while you're at it, please move ```Pager.__init__``` in the first line of the ```__init__``` method
 * your empty file doesn't need to be created in the setUp method as it's only used in your test. Please move the creation in your test, make it a local variable, and makes it follow twisted naming standard (filenameEmpty instead of self.filename_empty)
 * Use self.fail(msg) instead of ```raise Exception(msg)```

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="zectbumo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|zectbumo commented|
|-|-|

Replying to [therve](#note_3):
>  * while you're at it, please move ```Pager.__init__``` in the first line of the ```__init__``` method

I wish I could. The problem gets bigger if I try that. Pager.__init__ calls broker.registerPageProducer which eventually calls sendNextPage which needs a reference to self.chunks and self.chunks won't get initialize until after Pager.__init__.
The problem comes down to that Pager.__init__ is doing more than just initializing, it's also registering which causes wheels to start turning before we are done initializing.

>  * your empty file doesn't need to be created in the setUp method as it's only used in your test. Please move the creation in your test, make it a local variable, and makes it follow twisted naming standard (filenameEmpty instead of self.filename_empty)
>  * Use self.fail(msg) instead of ```raise Exception(msg)```

done

**Comment by twisted-trac** [comment]
|<img alt="zectbumo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|zectbumo set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set status to `closed`|
|-|-|

Fixed in r22475.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

(In [22482]) Revert r22475: stupid print statement.

Refs [#3023](https://github.com/twisted/twisted/issues/3023)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve commented|
|-|-|

(In [22485]) Branching to 'filepager-empty-file-3023'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve commented|
|-|-|

(In [22486]) Apply the patch, without the print.

Refs [#3023](https://github.com/twisted/twisted/issues/3023)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner<br>@therve set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @therve|
|-|-|

Humm okay.  I wish the PB code and tests were nicer. :(

But a bug is fixed, so this should go in.  Thanks, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve set status to `closed`|
|-|-|

(In [22557]) Merge filepager-empty-file-3023

Authors: zectbumo, therve
Reviewers: exarkun, therve
Fixes [#3023](https://github.com/twisted/twisted/issues/3023)

Make twisted.spread.util.FilePager works with empty files: previously, it
didn't callback correctly.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
